source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,203 @@
+};
+
+ScaledRangeTimerManager::ScaledRangeTimerManager(Dispatcher& dispatcher, float scale_factor,
+                                                 std::chrono::milliseconds minimum_duration)
+    : dispatcher_(dispatcher), minimum_duration_(minimum_duration), scale_factor_(scale_factor) {}
+
+RangeTimerPtr ScaledRangeTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<ScaledRangeTimer>(callback, *this);
+}
+
+void ScaledRangeTimerManager::setScaleFactor(float scale_factor) {
+  scale_factor_ = DurationScaleFactor(scale_factor);
+  for (auto& bucket : buckets_) {
+    if (!bucket.scaled_timers.empty()) {
+      bucket.updateTimer(*this);
+    }
+  }
+}
+
+MonotonicTime::duration
+ScaledRangeTimerManager::getBucketedDuration(MonotonicTime::duration max_duration) {
+  return MonotonicTime::duration(static_cast<MonotonicTime::duration::rep>(
+      std::pow(kBucketScaleFactor,
+               std::floor(std::log(max_duration.count()) / std::log(kBucketScaleFactor)))));
+}
+
+void ScaledRangeTimerManager::Bucket::updateTimer(ScaledRangeTimerManager& manager) {
+  if (scaled_timers.empty()) {
+    if (timer->enabled()) {
+      timer->disableTimer();
+    }
+  } else {
+    auto& entry = scaled_timers.front();
+    timer->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(
+        manager.scale_factor_.value() *
+        (entry.latest_trigger_time - manager.dispatcher_.timeSource().monotonicTime())));
+  }
+}
+
+ScaledRangeTimerManager::DurationScaleFactor::DurationScaleFactor(float value)
+    : value_(value < 0 ? 0 : value < 1 ? value : 1) {}
+
+float ScaledRangeTimerManager::DurationScaleFactor::value() const { return value_; }
+
+ScaledRangeTimerManager::BucketEnabledList::iterator
+ScaledRangeTimerManager::add(ScaledRangeTimer& timer, MonotonicTime::duration max_duration) {
+  max_duration = getBucketedDuration(max_duration);
+  Bucket& bucket = getBucket(max_duration);
+
+  bucket.scaled_timers.emplace_back(timer, max_duration + dispatcher_.timeSource().monotonicTime());
+  bucket.updateTimer(*this);
+  return --(bucket.scaled_timers.end());
+}
+
+void ScaledRangeTimerManager::disableActiveTimer(
+    MonotonicTime::duration max_duration, const BucketEnabledList::iterator& bucket_position) {
+  auto& bucket = getBucket(max_duration);
+  bucket.scaled_timers.erase(bucket_position);
+  bucket.updateTimer(*this);
+}
+
+void ScaledRangeTimerManager::onBucketTimer(int bucket_index) {
+  auto& bucket = buckets_[bucket_index];
+  ASSERT(!bucket.scaled_timers.empty());
+
+  auto it = bucket.scaled_timers.begin();
+  auto* timer = &it->timer;
+  bucket.scaled_timers.erase(it);
+
+  bucket.updateTimer(*this);
+
+  timer->trigger();
+}
+
+ScaledRangeTimerManager::Bucket&
+ScaledRangeTimerManager::getBucket(MonotonicTime::duration max_duration) {
+  const int index =
+      max_duration <= minimum_duration_
+          ? 0
+          : std::floor(std::log(max_duration.count()) / std::log(kBucketScaleFactor) -
```

Comment:
There may be a bug here related to the difference in units from max_duration and minimum_duration_.

This computation only makes sense if max_duration is in milliseconds.  You should accept a millisecond duration instead of duration.

Please add tests to cover the function that computes a bucket index from an input duration.

Reply:
Yeah this is a major bug, good catch. I have been trying to use MonotonicTime::duration everywhere in the manager, but I missed minimum_duration_. Now fixed.